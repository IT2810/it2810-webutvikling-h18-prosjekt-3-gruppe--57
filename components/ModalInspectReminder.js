import React from 'react';
import {
    ScrollView,
    Text,
    View,
    Image,
    Modal,
    TouchableHighlight,
} from 'react-native';
import {Kaede} from 'react-native-textinput-effects';
import { Overlay } from 'react-native-elements';
import { MapView } from 'expo';
import createStyles from '../styles/ModalInspectStyle.js';
import Storage from '../components/Storage';
import Score from '../components/Score';
import layout from '../constants/Layout';
import { Marker } from 'react-native-maps';

const styles = createStyles();

export default class ModalInspectReminder extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            modalVisible: false,
            id:false,
            img: null,
            textValue: null,
            result: null,
            overlayVisible: false,
            location: null,
            displayImage: 1,
            displayMap:'none',
        };
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        if (nextProps.modalVisible !== prevState.modalVisible || nextProps.id !== prevState.id) {
            return ({ modalVisible: nextProps.modalVisible,id:nextProps.id})
        }
        return (prevState);
    }

    _handleExit = () => {
        this.setState({ displayMap: 'none', img: null, displayImage: 1, location: null });
        this.props.refresh();
        this.props.setClose(false);
    }

    setImage(){
        Storage.getReminder(this.state.id).then((reminder) => {
            if(reminder.img != null){
                Score.updatePenalties(this.state.id, 'imgHint');
                this.setState({ img: reminder.img, displayImage: 200 });
            }else{
                alert("Sorry, no image was provided to this reminder");
            }
        });
    }

    setLocation(){
        Storage.getReminder(this.state.id).then((reminder)=>{
            if(reminder.location != null){
                Score.updatePenalties(this.state.id, "mapHint");
                this.setState({ location: reminder.location, displayMap:'flex'});
            }else{
                alert("No location was recorded for this reminder, you probably haven't given us permission to use the gps");
            }
        })
    }
    
    _updateScore = (failed=false) => Score.updateScore(this.state.id,failed).then(()=> this._handleExit()); 

    _updateAttempts = () => Score.incrementAttempts(this.state.id); 

    _compareInput = async (input) => {
        const reminder = await Storage.getReminder(this.state.id);
        if(reminder.reminder.toUpperCase() === input.toUpperCase()) return true;
        return false;
    }

    render() {
        return (
            <Modal
                animationType="slide"
                transparent={false}
                visible={this.state.modalVisible}
                onRequestClose={() => {
                    this.props.setClose(false);
                }}>
                <ScrollView style={{marginTop: 22}}>
                    <View style={styles.inputChooses}>
                        <Kaede onChangeText={(text) => { this.setState({ textValue: text }) }} label={'Remember your reminder?'}/>
                    </View>
                    <View style={styles.inputChooses}>
                        <TouchableHighlight
                            style={styles.button}
                            onPress={() => {
                                this.setImage()
                            }}>
                            <Text style={styles.modalText2}>View Image Hint</Text>
                        </TouchableHighlight>
                    </View>
                    <ScrollView style={{ height: this.state.displayImage }}>
                        <Image
                            style={{flex: 1,aspectRatio: 4 / 5}}
                            source={{ isStatic: true, uri: this.state.img }}
                        />
                    </ScrollView>
                    <View style={styles.inputChooses}>
                        <TouchableHighlight
                            style={styles.button}
                            onPress={() => {
                                this.setLocation()
                            }}>
                            <Text style={styles.modalText2}>View location Hint</Text>
                        </TouchableHighlight>
                    </View>
                    <MapView
                        style={{ flex: 1, width: layout.window.width, height: 200, display:this.state.displayMap }}
                        region={this.state.location}
                        zoomEnabled={false}
                        scrollEnabled={false}
                        minZoomLevel={14}
                    >
                        {this.state.location === null ? <View /> : <Marker
                            key={this.state.id}
                            title={"Hello"}
                            description={"Recorded location"}
                            coordinate={this.state.location}
                        /> }
                    </MapView>
                    <View style={styles.inputChooses}>
                        <TouchableHighlight
                            style={styles.buttonSave}
                            onPress={() => {
                                if(!this.state.textValue) alert("Input field is empty");
                                else{
                                    this._compareInput(this.state.textValue).then((res) => {
                                        if(res) this._updateScore();
                                        else {
                                            alert("Incorrect!");
                                            this._updateAttempts(); 
                                        }
                                    });
                                }         
                        }}>
                            <Text style={styles.modalText2}>Check</Text>
                        </TouchableHighlight>
                    </View>
                    <View style={styles.inputChooses}>
                        <TouchableHighlight
                            style={styles.buttonQuit}
                            onPress={() => {
                                this.setState({overlayVisible:true});
                            }}>
                            <Text style={styles.modalText2}>Give up</Text>
                        </TouchableHighlight>
                    </View>
                    <View style={styles.inputChooses}>
                        <TouchableHighlight
                            style={styles.buttonQuit}
                            onPress={() => this._handleExit()}
                            >
                            <Text style={styles.modalText2}>Go back</Text>
                        </TouchableHighlight>
                    </View>
                    <Overlay
                        isVisible={this.state.overlayVisible}
                        windowBackgroundColor="rgba(255, 255, 255, .8)"
                        height="auto"
                    >
                        <View>
                            <Text style={styles.modalText}>By proceeding you will lose 1000 points, continue?</Text>
                            <View style={styles.inputChooses}>
                                <TouchableHighlight
                                    style={styles.buttonSave}
                                    onPress={() => {
                                        this.setState({ overlayVisible: false });
                                        this._updateScore(true);
                                    }}>
                                    <Text style={styles.modalText2}>Continue</Text>
                                </TouchableHighlight>
                            </View>
                            <View style={styles.inputChooses}>
                                <TouchableHighlight
                                    style={styles.buttonQuit}
                                    onPress={() => {
                                        this.setState({ overlayVisible: false });
                                    }}>
                                    <Text style={styles.modalText2}>Take me back</Text>
                                </TouchableHighlight>
                            </View>
                        </View>
                    </Overlay>
                </ScrollView>
            </Modal>
        );
    }
}