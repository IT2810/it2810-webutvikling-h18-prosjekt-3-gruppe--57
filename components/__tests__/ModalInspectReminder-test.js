import React from "react";
import "react-native";
import renderer from "react-test-renderer";
import ModalInspectReminder from "../ModalInspectReminder";
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';

it('renders correctly', () => {
    const tree = renderer.create(<ModalInspectReminder/>).toJSON();
    expect(tree).toMatchSnapshot();
});


it('Snapshot of shallowRender.', ()=> {
    const renderer = new ShallowRenderer();
    renderer.render(<ModalInspectReminder/>);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
});


