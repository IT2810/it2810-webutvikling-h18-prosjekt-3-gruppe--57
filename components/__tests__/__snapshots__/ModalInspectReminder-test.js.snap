// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of shallowRender. 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={true}
>
  <ScrollViewMock
    style={
      Object {
        "marginTop": 22,
      }
    }
  >
    <Component
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <Kaede
        easing={[Function]}
        height={60}
        label="Remember your reminder?"
        onChangeText={[Function]}
      />
    </Component>
    <Component
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#DDDDDD",
            "padding": 10,
          }
        }
        underlayColor="black"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 17,
              "margin": 10,
              "textAlign": "center",
            }
          }
        >
          View Image Hint
        </Text>
      </TouchableHighlight>
    </Component>
    <ScrollViewMock
      style={
        Object {
          "height": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "isStatic": true,
            "uri": null,
          }
        }
        style={
          Object {
            "aspectRatio": 0.8,
            "flex": 1,
          }
        }
      />
    </ScrollViewMock>
    <Component
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#DDDDDD",
            "padding": 10,
          }
        }
        underlayColor="black"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 17,
              "margin": 10,
              "textAlign": "center",
            }
          }
        >
          View location Hint
        </Text>
      </TouchableHighlight>
    </Component>
    <MapView
      minZoomLevel={14}
      region={null}
      scrollEnabled={false}
      style={
        Object {
          "display": "none",
          "flex": 1,
          "height": 200,
          "width": 750,
        }
      }
      zoomEnabled={false}
    >
      <Component />
    </MapView>
    <Component
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#17cf94",
            "padding": 10,
          }
        }
        underlayColor="black"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 17,
              "margin": 10,
              "textAlign": "center",
            }
          }
        >
          Check
        </Text>
      </TouchableHighlight>
    </Component>
    <Component
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F2686B",
            "padding": 10,
          }
        }
        underlayColor="black"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 17,
              "margin": 10,
              "textAlign": "center",
            }
          }
        >
          Give up
        </Text>
      </TouchableHighlight>
    </Component>
    <Component
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F2686B",
            "padding": 10,
          }
        }
        underlayColor="black"
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 17,
              "margin": 10,
              "textAlign": "center",
            }
          }
        >
          Go back
        </Text>
      </TouchableHighlight>
    </Component>
    <Overlay
      borderRadius={3}
      fullScreen={false}
      height="auto"
      isVisible={false}
      overlayBackgroundColor="white"
      width={670}
      windowBackgroundColor="rgba(255, 255, 255, .8)"
    >
      <Component>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#000",
              "fontSize": 27,
              "margin": 10,
              "textAlign": "center",
            }
          }
        >
          By proceeding you will lose 1000 points, continue?
        </Text>
        <Component
          style={
            Object {
              "paddingVertical": 10,
            }
          }
        >
          <TouchableHighlight
            activeOpacity={0.85}
            delayPressOut={100}
            onPress={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#17cf94",
                "padding": 10,
              }
            }
            underlayColor="black"
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000",
                  "fontSize": 17,
                  "margin": 10,
                  "textAlign": "center",
                }
              }
            >
              Continue
            </Text>
          </TouchableHighlight>
        </Component>
        <Component
          style={
            Object {
              "paddingVertical": 10,
            }
          }
        >
          <TouchableHighlight
            activeOpacity={0.85}
            delayPressOut={100}
            onPress={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#F2686B",
                "padding": 10,
              }
            }
            underlayColor="black"
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000",
                  "fontSize": 17,
                  "margin": 10,
                  "textAlign": "center",
                }
              }
            >
              Take me back
            </Text>
          </TouchableHighlight>
        </Component>
      </Component>
    </Overlay>
  </ScrollViewMock>
</Component>
`;

exports[`renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={true}
/>
`;
